miniapps/meshing/reflector.cpp

```diff
@@ -0,0 +1,1061 @@
+      r2o[copiedElem] = e;
+
+      // Add the new reflected element
+      Array<int> rvert(elvert.Size());
+      for (int i=0; i<elvert.Size(); ++i)
+      {
+         const int v = elvert[i];
+         rvert[i] = (v2r[v] == -1) ? v : v2r[v];
+      }
+
+      const int newElem = builder.AddElement(rvert, onPlane);
+      r2o[newElem] = -1 - e;
+   }
+
+   Mesh *reflected = builder.mesh;
+
+   // Set attributes
+   MFEM_VERIFY((int) r2o.size() == reflected->GetNE(), "");
+   for (int i = 0; i < (int) r2o.size(); ++i)
+   {
+      const int e = (r2o[i] >= 0) ? r2o[i] : -1 - r2o[i];
+      reflected->SetAttribute(i, mesh.GetAttribute(e));
+   }
+
+   // In order to set boundary attributes, first set a map from original mesh
+   // boundary elements to reflected mesh boundary elements, by using the vertex
+   // map v2r. Note that for v < mesh.GetNV(), vertex v of `mesh` coincides with
+   // vertex v of `reflected`, and if that vertex is not in the reflection
+   // plane, v2r[v] >= mesh.GetNV() is the index of the vertex in `reflected`
+   // that is its reflection.
+
+   // Identify each quadrilateral boundary element with the unique pair of
+   // vertices (v1, v2) such that v1 is the minimum vertex index in the
+   // quadrilateral, and v2 is diagonally opposite v1.
+
+   std::map<std::pair<int, int>, int> mapBE;
+   for (int i=0; i<mesh.GetNBE(); ++i)
+   {
+      const Element *be = mesh.GetBdrElement(i);
+      Array<int> v;
+      be->GetVertices(v);
+      MFEM_VERIFY(v.Size() == 4, "Boundary elements must be quadrilateral");
+
+      const int v1 = v.Min();
+      int v1i = -1;
+      for (int j=0; j<v.Size(); ++j)
+      {
+         if (v[j] == v1)
+         {
+            v1i = j;
+         }
+      }
+
+      const int v2 = v[(v1i + 2) % 4];
+
+      mapBE[std::pair<int, int>(v1, v2)] = i;
+
+      // Find the indices of vertices in `reflected` of the reflected quadrilateral.
+      Array<int> rv(4);
+      int rv1 = -1;  // Find the minimum reflected vertex index.
+      int rv1i = -1;
+      bool inPlane = true;
+      for (int j=0; j<v.Size(); ++j)
+      {
+         rv[j] = (v2r[v[j]] == -1) ? v[j] : v2r[v[j]];
+
+         if (v2r[v[j]] != -1)
+         {
+            inPlane = false;
+         }
+
+         if (rv1 == -1 || rv[j] < rv1)
+         {
+            rv1 = rv[j];
+            rv1i = j;
+         }
+      }
+
+      // Note that in-plane boundary elements are skipped.
+      if (!inPlane)
+      {
+         const int rv2 = rv[(rv1i + 2) % 4];
+
+         mapBE[std::pair<int, int>(rv1, rv2)] = i;
+
+         const Geometry::Type orig_geom = mesh.GetBdrElementBaseGeometry(i);
+         Element *rbe = reflected->NewElement(orig_geom);
+         rbe->SetVertices(v);
+         reflected->AddBdrElement(rbe);
+
+         rbe = reflected->NewElement(orig_geom);
+         rbe->SetVertices(rv);
+         reflected->AddBdrElement(rbe);
+      }
+   }
+
+   for (int i=0; i<reflected->GetNBE(); ++i)
```

Comment:
When running the sample runs I saw a very familiar warning:
```Boundary elements with wrong orientation: 20 / 40 (fixed)```

Clearly we have an algorithm which can fix this so we could just make that part of the reflection process so that users don't get this warning. Ideally we could reorder the reflected boundary element vertices (similar to what is done for the reflected hexes) to avoid this.

Reply:
This is fixed now with 1 line, to swap the 0 and 2 vertices of each reflected quadrilateral boundary element.